Mathlib/Geometry/Manifold/IntegralCurve.lean

```diff
@@ -0,0 +1,767 @@
+
+  This is stated for manifolds without boundary for simplicity. We actually only need to assume that
+  the images of `γ` and `γ'` lie in the interior of the manifold. -/
+lemma isIntegralCurveOn_piecewise [BoundarylessManifold I M]
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M)))
+    {a b a' b' : ℝ} (hγ : IsIntegralCurveOn γ v (Ioo a b))
+    (hγ' : IsIntegralCurveOn γ' v (Ioo a' b')) {t₀ : ℝ}
+    (ht₀ : t₀ ∈ Ioo a b ∩ Ioo a' b') (h : γ t₀ = γ' t₀) :
+    IsIntegralCurveOn (piecewise (Ioo a b) γ γ') v (Ioo a b ∪ Ioo a' b') := by
+  intros t ht
+  by_cases hmem : t ∈ Ioo a b
+  · rw [piecewise, if_pos hmem]
+    apply (hγ t hmem).congr_of_eventuallyEq
+    rw [Filter.eventuallyEq_iff_exists_mem]
+    refine ⟨Ioo a b, isOpen_Ioo.mem_nhds hmem, ?_⟩
+    intros t' ht'
+    rw [piecewise, if_pos ht']
+  · rw [mem_union, or_iff_not_imp_left] at ht
+    have hmem' := ht hmem
+    rw [piecewise, if_neg hmem]
+    apply (hγ' t hmem').congr_of_eventuallyEq
+    rw [Filter.eventuallyEq_iff_exists_mem]
+    refine ⟨Ioo a' b', isOpen_Ioo.mem_nhds hmem', piecewise_eqOn_symm hv hγ hγ' ht₀ h⟩
+
+/-- If there exists `ε > 0` such that the local integral curve at each point `x : M` is defined at
+  least on an open interval `Ioo (-ε) ε`, then every point on `M` has a global integral
+  curve passing through it.
+
+  See Lemma 9.15, Lee -/
+lemma exists_isIntegralCurve_of_isIntegralCurveOn [BoundarylessManifold I M]
+    {v : (x : M) → TangentSpace I x}
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M)))
+    {ε : ℝ} (hε : 0 < ε) (h : ∀ x : M, ∃ γ : ℝ → M, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-ε) ε))
+    (x : M) : ∃ γ : ℝ → M, γ 0 = x ∧ IsIntegralCurve γ v := by
+  have hnon : Set.Nonempty {a | ∃ γ, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-a) a)} := ⟨ε, h x⟩
+  by_cases hbdd : BddAbove {a | ∃ γ, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-a) a)}
+  · set asup := sSup {a | ∃ γ, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-a) a)} with hasup
+    -- we will obtain two integral curves, one centred at some `t₀ > 0` with
+    -- `0 ≤ asup - ε < t₀ < asup`; let `t₀ = asup - ε / 2`
+    -- another centred at 0 with domain up to `a ∈ S` with `t₀ < a < asup`
+    obtain ⟨a, ha, hlt⟩ := Real.add_neg_lt_sSup hnon (ε := - (ε / 2))
+      (by rw [neg_lt, neg_zero]; exact half_pos hε)
+    have hale : a ≤ asup := le_csSup hbdd ha
+    rw [mem_setOf] at ha
+    rw [← hasup, ← sub_eq_add_neg] at hlt
+
+    have hεle : ε ≤ asup := le_csSup hbdd (h x)
+
+    -- integral curve defined on `Ioo (-a) a`
+    obtain ⟨γ, h0, hγ⟩ := ha
+    -- integral curve starting at `-(asup - ε / 2)` with radius `ε`
+    obtain ⟨γ1_aux, h1_aux, hγ1_aux⟩ := h (γ (-(asup - ε / 2)))
+    let γ1 := γ1_aux ∘ (· + (asup - ε / 2))
+    have heq1 : γ1 (-(asup - ε / 2)) = γ (-(asup - ε / 2)) := by simp [h1_aux]
+    have hγ1 : IsIntegralCurveOn γ1 v (Ioo (-(asup + ε / 2)) (-(asup - ε))) := by
```

Comment:
There's quite a bit of code duplication in here - I'm sure this can be refactored (and perhaps be extracted into a separate lemma).